Handle the not in reference keys, simple way.

Valc opened the pull request:
@infograf768

After playing a bit with different alternatives, here I send you the PR that I liked the most as it was: simple and practical.

I think it is a task that requires being careful to select what to delete and what not, and I have avoided putting a button type "Select all"

I have also changed the output a bit so that everything is more in line and you can be sure that the selected box corresponds to the key.

And not much else :)

Here some screenshots

![nir1](https://user-images.githubusercontent.com/417247/134815748-4b8b68ad-d4c2-42a6-b554-56dc2f9437f4.png)

![nir2](https://user-images.githubusercontent.com/417247/134815978-476175b9-ef8a-43b3-83d2-97c5e4ae005b.png)

infograf768 commented:
This PR does not work here.
I do not get the checkbox and the display has not changed.
I get the message although I did not (could not) check a box.
<img width="962" alt="Screen Shot 2021-09-27 at 10 11 29" src="https://user-images.githubusercontent.com/869724/134870004-c0156502-46da-4d8b-b547-c6eea573f9a1.png">
<img width="954" alt="Screen Shot 2021-09-27 at 10 13 11" src="https://user-images.githubusercontent.com/869724/134870085-049df504-bc71-4fbd-81e2-8e6ce3840936.png">

There are other parts I would like to discuss but first this has to work.

infograf768 commented:
@valc
looks like part of the problem is that `$istranslation` is = 0 in keyfield

infograf768 commented:
Also, the pr breaks the filters.

infograf768 commented on administrator/components/com_localise/tmpl/translation/edit.php:
```diff
@@ -68,31 +68,27 @@
-	$keystofilter	= array ($this->item->$filter);
-	$tabchoised		= 'strings';
+	$filter      = $posted['select']['keystatus'];
+	$keystofilte = array ($this->item->$filter);
```
`keystofilter`

Valc commented:
@infograf768 

I will need to investigate the matter a little more in depth because this goes beyond the errors that I introduced in my last PR

If you fix the errors you found and remove the "istranslation" check that I added to the last one, it should work for you as seen on the screens.

But it does so from the wrong logic. That place where I put the PR is for when the language "en-GB" is selected

```
			if ($status= 'extra')
			{
				if ($istranslation)
				{
```

Maybe I introduced it years ago (could be) but not with this PR. With this one bugged the PR was working also on not en-GB tags:

`if ($status= 'extra')` 

So, is bugged and i need take a look about how to solve from all cases and then insert the PR code at right location.

Valc commented:
I think i have catched the main problem at Translation model "preprocessForm" fuction:

```
if (!$newstrings)
							{
								$newstrings = true;
								$form->load($addform, false);
								$section = 'COM_LOCALISE_TEXT_TRANSLATION_NOTINREFERENCE';
								$addform = new \SimpleXMLElement('<form />');
								$group   = $addform->addChild('fields');
								$group->addAttribute('name', 'strings');
								$fieldset = $group->addChild('fieldset');
								$fieldset->addAttribute('name', $section);
								$fieldset->addAttribute('label', $section);
								$field->addAttribute('istranslation', $istranslation);
							}

							$field   = $fieldset->addChild('field');
							$status  = 'extra';
							$default = $string;
							$label   = '<strong>' . $key . '</strong>';
							$field->addAttribute('status', $status);
							$field->addAttribute('description', $string);
							$field->addAttribute('istranslation', $istranslation);
```

It was skiping add the "istranslation attribute" when extra cases. So, all "if is translation" at KeyField was crashing always with a false value when extra cases matches.

Now i will go faster send the new PR :)

Valc commented:
@infograf768

Now I have gone over everything a bit.

**Apart from what has already been said:**

- The JS was breaking in the "en-GB" view because the filters are not included there.
- The "Revert all" and "Revert" buttons were also not in the en-GB view.
- I think I have tried all cases of both text changes and deleting extras in en-GB and xx-XX and I think now everything is going as expected.
- I have made an effort so that the "labels" and the "imput" of each case are easier to follow and modify from the "KeyField"
- Now the output screens are no longer the same as the ones you sent at first, but here is dinner time and maybe later add the new ones :D.

If no more bugs to solve.. suggestions time :)

Valc commented:
**Translation cases**
![nir3](https://user-images.githubusercontent.com/417247/135346437-7c60e57d-72f8-4694-a96a-d0beb3fb03de.png)
![nir4](https://user-images.githubusercontent.com/417247/135346440-657fd2f3-c3c6-4e66-84c7-166c088fd670.png)
![nir5](https://user-images.githubusercontent.com/417247/135346441-3aea2581-0388-4f8c-b7d7-1cb8cc4cae75.png)

**en-GB**
![nir6](https://user-images.githubusercontent.com/417247/135346443-822eb670-c5ee-49b9-bb4f-fa0084c1b61c.png)

infograf768 commented:
I am testing, but first please explain what you mean by "Grammar". I know it was hardcoded before but, as I never had the case, I did not mind. Now that it is a lang string, I wonder what you mean.

infograf768 commented:
Much better :)
My first tests results.

1. No alert when some strings are checked when saving. We should be able to cancel the save or save and close.
2. If we have multiple strings in `Not in reference` and we only select some of them to delete, the resulting message does not take this situation into consideration. It only displays a message informing that some strings have been deleted.
I.e. we may need these two messages in this case
`COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF`
`COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF_OMITED` // btw change to OMITTED

In any case the string 
`COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF="The keys that you selected because they no longer exist in the reference language have been deleted."`
can be changed to
`COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF="The strings selected in the \"Not in reference\" section have been deleted."`

---Note: please change the word keys to strings generally speaking in the ini file. 

3. `[Not in Ref]` in red is redundant. I rather suggest to add a general info when displaying the not in reference section.
For example:

<img width="1112" alt="Screen Shot 2021-09-30 at 10 40 12" src="https://user-images.githubusercontent.com/869724/135418466-e68f3aa1-b68e-4eeb-856b-de7b2663b24b.png">

code to take off the red text:
replace line 1108 in TranslationModel by
`$label   = '<br><strong>' . $key . '</strong>';`
 and take off the string `COM_LOCALISE_NOTINREF_KEY="Not in ref"`

Change ini string to
`COM_LOCALISE_TOOLTIP_TRANSLATION_EXTRA_KEYS_IN_TRANSLATION="If you do not want to keep a string in the translation, you should check the checkbox to remove it on 'Save'.<br> Some strings may not be in reference anymore.<br>Some strings may be specific plural forms that you need to keep."`

Suggestion of code to display the info
```
									foreach ($sections as $name => $fieldSet) :
										echo HTMLHelper::_('bootstrap.addSlide', 'localise-translation-sliders', Text::_($fieldSet->label), 'collapse' . $i++);

										if ($fieldSet->label == "COM_LOCALISE_TEXT_TRANSLATION_NOTINREFERENCE") : ?>
											<div class="alert alert-info">
												<span class="fas fa-info-circle info-line" aria-hidden="true"</span><span class="sr-only"><?php echo Text::_('INFO'); ?></span>
												<?php echo Text::_('COM_LOCALISE_TOOLTIP_TRANSLATION_EXTRA_KEYS_IN_TRANSLATION'); ?>
											</div>
										<?php endif; ?>
										<ul class="adminformlist">
```

Will look again later

Valc commented:
@infograf768 
> I am testing, but first please explain what you mean by "Grammar". I know it was hardcoded before but, as I never had the case, I did not mind. Now that it is a lang string, I wonder what you mean.

The text changes come to us from the en-GB language so that in our translation "xx-XX" we apply the same differences on the translation into our language.

Imagine that in the next release of Joomla, on the same file, there have been` 7  keys `with changed text in "`en-GB`"

`Of those 7 cases, there is 1 that consists of correcting:
`
This step is to be om**i**ted by This step is to be omi**tt**ed

"omi**t**ed" is grammatically incorrect in en-GB, and then you mark that one as "Grammar case" from editing the en-GB file.

When the news is shown to me in that same file, but in my translation into Spanish, `I will only receive 6 cases` per changed text to review: those 6 that are not really "en-GB grammatical cases", "typing errors" "Typo cases", or whatever it is called in English.

Of course, in my case i only use com_localise at our translation site to translate from en-GB to the es-ES and no more languages: it can to be useful when Joomla is "under grammatical review" (some times happens)

If the case were that I want to translate from en-GB to en-US, then it would not be very recommended to use this functionality.

infograf768 commented:
> `omi**t**ed` is grammatically incorrect in en-GB, and then you mark that one as "Grammar case" from editing the en-GB file.

Not sure I understand. How do you "mark" a string as `Grammar Case`? Please provide a screenshot or, better an animated gif.

Valc commented:
From languages view select any client and language "en-GB".

From the translations list, search a file with "Alert icon" and info about "Grammar cases" detected.

Edit that file.

infograf768 commented:
@valc
Lost in translation. Please post an animated gif. I don't see anywhere `Changed texts or Grammar cases`.
But i see some weird behavior
![grammar](https://user-images.githubusercontent.com/869724/135577910-64541f95-5188-4471-b8f6-9b7998a2be17.gif)
.

infograf768 commented:
Please test saving a file which does not contain any "Not in reference" string. See the Notice. It is still displaying "You are maintaining..." 
This is due to `elseif (empty($notinref[0])) etc.` coming as a sub of `if (!empty($strings) && !empty($notinref) && $istranslation)` Weird.

Please use Text::plural in the model for the Notices and modify strings values. For example:

```
COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF_1="%1$s selected string has been deleted as it was no more present in the reference language."
COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF_MORE="%1$s selected strings have been deleted as they were no more present in the reference language."

COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF_OMITTED_1="You are maintaining %1$s string that does not exist in the reference language."
COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF_OMITTED_MORE="You are maintaining %1$s strings that do not exist in the reference language."
```

I do not see any use for
`COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF_ALL="Now you are not maintaining strings that do not exist in the reference language."`


Concerning the alert, I would make it general and add it to the Save alert in the view. I guess that would be the simplest way.

Valc commented:
@infograf768
> Lost in translation. Please post an animated gif. I don't see anywhere `Changed texts or Grammar cases`.

With what program to make animated Gifs? Is there for Linux?

You are in the right place, it is that I have explained myself very badly.

Actualy no text changes comming between the last 2 Joomla releases. So, Select "4.0.0" from "Options" as "Source reference":

![grammar1](https://user-images.githubusercontent.com/417247/135648881-073c37fa-be95-44d6-9566-23e1d8727f8d.png)

1. Apply the next filters at Translations view to get only incomplete content at admin client for the en-GB language
2. Search at the tables files list moving the muose over the "alert icon" informing about changes in develop, if present.

![grammar2](https://user-images.githubusercontent.com/417247/135648884-d15d6d6c-167d-42da-9c9e-485dabd6a2d1.png)

`com_actionlogs.ini` explains "0 Grammar cases of 2 changed text", so: edit that file:

1. For testing purposes only check as "grammar case" the key `COM_ACTIONLOGS_N_ITEMS_DELETED`
2. Save and close

![grammar3](https://user-images.githubusercontent.com/417247/135648888-3b49a7e1-f861-4242-ae62-93e918030292.png)

1. Now select the French language from Translations view
2. In your translation and for the same file, you will receive the notification that there is only 1 text change.
3. If you edit the fr-FR com_actionlogs.ini file no "COM_ACTIONLOGS_N_ITEMS_DELETED" key as changed text is present

![grammar4](https://user-images.githubusercontent.com/417247/135654585-0935f9f1-68f2-40cb-911f-0b76578e4269.png)


That was for testing purposes only, because it was not really a grammatical case of en-GB

1. Select again "en-GB", edir the same file and uncheck as "Grammar case" the key "COM_ACTIONLOGS_N_ITEMS_DELETED"
2. Save and close.
3. Select the fr-FR language and then:

![grammar5](https://user-images.githubusercontent.com/417247/135654637-4c383310-5789-4888-8691-b57410a65bc3.png)

Now is informing about 2 changed text to revise at fr-FR language, but with a bugged result for the revised ones at fr-FR (i need a take a look about why due before was working fine)

Valc commented:
> Please test saving a file which does not contain any "Not in reference" string. See the Notice. It is still displaying "You are maintaining..." This is due to `elseif (empty($notinref[0])) etc.` coming as a sub of `if (!empty($strings) && !empty($notinref) && $istranslation)` Weird.
> 
> Please use Text::plural in the model for the Notices and modify strings values. For example:
> 
> ```
> COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF_1="%1$s selected string has been deleted as it was no more present in the reference language."
> COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF_MORE="%1$s selected strings have been deleted as they were no more present in the reference language."
> 
> COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF_OMITTED_1="You are maintaining %1$s string that does not exist in the reference language."
> COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF_OMITTED_MORE="You are maintaining %1$s strings that do not exist in the reference language."
> ```
> 
> I do not see any use for `COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF_ALL="Now you are not maintaining strings that do not exist in the reference language."`
> 
> Concerning the alert, I would make it general and add it to the Save alert in the view. I guess that would be the simplest way.

I will take a look ASAP

Valc commented:
@infograf768 
> Please test saving a file which does not contain any "Not in reference" string

Here i am not able to reproduce the issue. Tested with `fr-FR com_admin.ini` file at admin client, with no cases of "Not in reference" in this moment.

Please, let me know what file are you testing for the wrong result.

> I do not see any use for COM_LOCALISE_NOTICE_TRANSLATION_DELETE_NOTINREF_ALL="Now you are not maintaining strings that do not exist in the reference language."

If we have X cases and we delete all cases at time and `we have not a "You have deleted XX of the YY cases" info message`, i think than then the "ALL" have sense due is noticed about how many keys has been deleted and we are also informed than there are no more "Not in reference" cases to handle in that moment.

About alert message i will try tomorrow to add the code at the PR, the other one i think is added :)

Valc commented:
@infograf768 
https://github.com/infograf768/j4localise/pull/35#issuecomment-932373096
> Now is informing about 2 changed text to revise at fr-FR language, but with a bugged result for the revised ones (i need take a look about why due before was working fine)

After searching in depth and using different lines of code to do the same, I have realized that what is happening has nothing to do with the wrong data being stored in the db or that the code does not know how to count the cases correctly en-GB or xx-XX: I think it does it perfectly.

I think the problem comes from not having "read-only" protected the "Grammar cases" since their "en-GB" edition. I think that is where the matter goes, tomorrow I will try to continue investigating.

infograf768 commented:
Honestly, @valc, I would take off the grammar case stuff. It looks like a very limited and confusing "feature". "Modified value" is enough imho.

Will test again
Note: I am using LICEcap on Macintosh
Here is a link to similar software for Linux: https://alternativeto.net/software/licecap/?platform=linux

infograf768 commented:
>  Please test saving a file which does not contain any "Not in reference" string

> Here i am not able to reproduce the issue. Tested with fr-FR com_admin.ini file at admin client, without cases of "Not in reference" in this moment.

IMPORTANT:
com_admin.ini in French has a lot of not in reference strings in 4.0.4-dev compared to 4.0.3 but the Info is wrong as well as the display in Not in Reference. BUG!! and not only after this patch.
<img width="1375" alt="Screen Shot 2021-10-02 at 10 51 09" src="https://user-images.githubusercontent.com/869724/135709830-a56dda17-9f98-4513-8a89-ecc125353482.png">

These are the obsolete help strings which are totally replaced in 4.0.4-dev. These should be deletable.

Here is a test with actionlogs showing the "remaining issue"
![remaining](https://user-images.githubusercontent.com/869724/135709640-a89224be-a990-4f1f-9236-4c80150150d8.gif)

Valc commented:
> Here is a test with actionlogs showing the "maintaining" notice

Confirmed. With that file i also get a bugged reply message. I will take a look.

>IMPORTANT:
com_admin.ini in French has a lot of not in reference strings in 4.0.4-dev compared to 4.0.3 but the Info is wrong as well as the display in Not in Reference. BUG!! and not only after this patch.

Please, try to "reinstall" from Joomla languages installer  your language package and then take a look again about "not in ref" on that file.

> EDIT: Maybe this bug is due to the new : in the new strings constant?

I think you have correctly solved that issue with your patch. Try reinstalling your language package first.

With the other PT of "remove extras" I got excited and left everything that was left removed and now, as usual, they no longer come out as extras. To be able to see them again I would have to reinstall my language pack from the installer.

 Is your case?


> Honestly, @Valc, I would take off the grammar case stuff. It looks like a very limited and confusing "feature". "Modified value" is enough imho.
Totally agree.

Valc commented:
After reinstall your language pack or the es-ES (no matter), same file but raw mode edition:

![bug00](https://user-images.githubusercontent.com/417247/135725160-fa8f1bcb-703c-4bf5-94c5-2ae7813bee18.png)
Save and close Raw edition

Same file edition and then "Save" (we are ar not raw mode now)

Same file edition again in raw mode

![bug00](https://user-images.githubusercontent.com/417247/135725279-8907976c-e62e-43bd-bad7-aa57bbb34723.png)
![bug1](https://user-images.githubusercontent.com/417247/135725281-45d62cfb-ec5a-44ca-a46c-7fa02e222a0c.png)

Valc commented:
@infograf768
I got it 8)

It was related with the stream end of file handler. the bug is only present when the file have not an extra empty line at the end of the file to translate.

Now i have simply moved the "$line = $stream->gets();" at the top of the stream code start creating the language file output:
```
			while (!$stream->eof())
			{
			// Mounting the language file in this way will help to avoid save files with errors at the content.
				$line = $stream->gets();
				// Blank lines
				if (preg_match('/^\s*$/', $line))

And then the last line of the file is also always under revision.

If is mandatory add an extra empty line at the end of the language file, let me know due now is not added.
```

Valc commented:
Added the message alert.
I can not share the screenshot due some lines on the top of my laptop screen are a bit damaged and the alert message is not easy to read.

infograf768 commented:
I had to make a new installation of J4-dev to at last get a correct notinref delete for admin.ini.
before I test again with your last changes and correct comments and strings English, please correct the new save alert in the view.

```
if ($complete === 1)
			{
				$message = 'COM_LOCALISE_CONFIRM_TRANSLATION_SAVE';

				if ($has_notinref && $istranslation)
				{
					$message = 'COM_LOCALISE_CONFIRM_TRANSLATION_SAVE_AND_NOTINREF';
				}
```
The notinref message should be independent from the $complete variable.

Valc commented:
@infograf768 
> The notinref message should be independent from the $complete variable.

Before continuing I have some questions:

1. The alert message 'Not in ref' also has to be conditioned to a parameter configurable from "Options" such as with "Complete"
2. The way I understand it to apply your suggestion is than they are going to have to jump as two alarm messages: first one and if not canceled then the other, is that what you propose?

infograf768 commented:
> The alert message 'Not in ref' also has to be conditioned to a parameter configurable from "Options" such as with "Complete"?

Not necessary imho.

> The way I understand it to apply your suggestion is than they are going to require be showed as two alarm messages: first one and if not canceled then the other, is that what you propose?

Not really. We do not need any more the complete message in that case.

maybe something like (I did not test):
```
			if ($complete === 1)
			{
				$message = 'COM_LOCALISE_CONFIRM_TRANSLATION_SAVE';

				if ($has_notinref && $istranslation)
				{
					$message = 'COM_LOCALISE_CONFIRM_TRANSLATION_SAVE_AND_NOTINREF';
				}

				$toolbar->confirmButton('apply')
					->text('JAPPLY')
					->message($message)
					->task('translation.apply');

				$toolbar->confirmButton('save')
					->text('JSAVE')
					->message($message)
					->task('translation.save');
			}
			elseif (!$complete && $has_notinref && $istranslation)
			{
				$message = 'COM_LOCALISE_CONFIRM_TRANSLATION_SAVE_AND_NOTINREF';
				
				$toolbar->confirmButton('apply')
					->text('JAPPLY')
					->message($message)
					->task('translation.apply');

				$toolbar->confirmButton('save')
					->text('JSAVE')
					->message($message)
					->task('translation.save');
			}
			else
			{
				ToolbarHelper::apply('translation.apply');
				ToolbarHelper::save('translation.save');
			}
```
I will look anyway at the lang strings later.

infograf768 commented:
Found another oddity:
When editing a com_localise.ini file (which is not a core file) I get an error in the view
```
Notice: Undefined index: extra_keys in /Applications/MAMP/htdocs/sample/joomla40/administrator/components/com_localise/View/Translation/HtmlView.php on line 83

Notice: Trying to access array offset on value of type null in /Applications/MAMP/htdocs/sample/joomla40/administrator/components/com_localise/View/Translation/HtmlView.php on line 83

```
I guess we have to test if `($this->get('Item')->developdata['extra_keys']` exists before defining the new variable.

Valc commented:
> Found another oddity: When editing a com_localise.ini file (which is not a core file) I get an error in the view
> 
> ```
> Notice: Undefined index: extra_keys in /Applications/MAMP/htdocs/sample/joomla40/administrator/components/com_localise/View/Translation/HtmlView.php on line 83
> 
> Notice: Trying to access array offset on value of type null in /Applications/MAMP/htdocs/sample/joomla40/administrator/components/com_localise/View/Translation/HtmlView.php on line 83
> ```
> 
> I guess we have to test if `($this->get('Item')->developdata['extra_keys']` exists before defining the new variable.

Utch, yes!
 Only core files have develop from Github, son also is required the condition  if "have delelop"

I will take a look here and at the alarm suggestion such as you exlain

Valc commented:
I have been trying to apply your suggestions but not all cases are under control with the new code lines. And at the end i think is required combine both messages on same alert message.

Maybe you can found a better combo, please, make your tests and let me know.

I think the "Translation completed" parameter can become "Show reminder alerts" as configurable param and then can keep more than one "message" to show as reminder alerts

If the user does not like get "reminders" no one will be showed and if he have set it as default to " show" then he can be noticed about 1 or more reminders on same alert box message.

infograf768 commented:
OK for a unique "Display alerts"

Valc commented:
Done.
Job time. I will read next suggestions later.

infograf768 commented on administrator/components/com_localise/Helper/LocaliseHelper.php:
```diff
@@ -1865,6 +1865,7 @@ public static function getDevelopchanges($info = array(), $refsections = array()
 						}
 						else
 						{
+
```
delete line

infograf768 commented on administrator/components/com_localise/View/Translation/HtmlView.php:
```diff
@@ -103,22 +110,28 @@ protected function addToolbar()
+
+				if ($has_notinref && $istranslation)
+				{
+					$message .= '\\n' . Text::_('COM_LOCALISE_CONFIRM_TRANSLATION_SAVE_NOTINREF');
```
add a return tmake message clearer
`$message .= '\\n\\n' . Text::_('COM_LOCALISE_CONFIRM_TRANSLATION_SAVE_NOTINREF');

infograf768 commented on administrator/components/com_localise/language/en-GB/com_localise.ini:
```diff
@@ -188,10 +188,13 @@ COM_LOCALISE_CANNOT_REMOVE_DEFAULT_LANGUAGE="Localise cannot remove default lang
 
 COM_LOCALISE_ACCESS_TRANSLATION_EDIT="Translate"
 COM_LOCALISE_ACCESS_TRANSLATION_EDIT_DESC="Translate description"
+COM_LOCALISE_CHECKBOX_TRANSLATION_GRAMMAR_CASE="Grammar case"
```
Is that one still used? I see also the code
```
if ($istranslation)
			{
				$title = Text::_('COM_LOCALISE_REVISED');
				$tip   = $title;
			}
			else
			{
				$title = Text::_('COM_LOCALISE_CHECKBOX_TRANSLATION_GRAMMAR_CASE');
				$tip   = Text::_('COM_LOCALISE_TOOLTIP_TRANSLATION_GRAMMAR_CASE');
			}
```

How to test?
If it is used, the English "Grammar case" is confusing.

infograf768 replied:
I found how to test.
The tip is almost unreadable which means it should be displayed in another way and we definitely should use `Modified` instead of the wording "Grammar case" as this is not applicable to the situation.
![Screen Shot 2021-10-04 at 12 05 16](https://user-images.githubusercontent.com/869724/135833012-db2f2f7d-358b-4ffd-9836-0cbd3de3fc7c.png)

Valc replied:
Nice idea, i will take a look.

infograf768 commented on administrator/components/com_localise/language/en-GB/com_localise.ini:
```diff
@@ -188,10 +188,13 @@ COM_LOCALISE_CANNOT_REMOVE_DEFAULT_LANGUAGE="Localise cannot remove default lang
 COM_LOCALISE_ERROR_TRANSLATION_FILESAVE="Error saving translation file %s"
 COM_LOCALISE_FIELDSET_TRANSLATION_DETAIL="Details"
-COM_LOCALISE_FIELDSET_TRANSLATION_DETAIL_DESC="Values ​​of inactive fields are set in general settings of Localise component."
+COM_LOCALISE_FIELDSET_TRANSLATION_DETAIL_DESC="Values of inactive fields are set in general settings of Localise component."
```
Suggest
`COM_LOCALISE_FIELDSET_TRANSLATION_DETAIL_DESC="Values of inactive fields are set in the component options"

Valc replied:
I will apply the suggestion but no string changed from my part there. It was due from the Github edition of the file it was showing a "red dot" there, between the words, so.. odd dot.. i delete  the dot XD.

infograf768 commented on administrator/components/com_localise/language/en-GB/com_localise.ini:
```diff
@@ -232,13 +240,20 @@ COM_LOCALISE_TEXT_TRANSLATION_TRANSLATED="Translated"
 COM_LOCALISE_TEXT_TRANSLATION_UNTRANSLATED="Untranslated"
 COM_LOCALISE_TITLE_TRANSLATION="Translation"
-COM_LOCALISE_TOOLTIP_TRANSLATION_INSERT="<strong>Insert</strong><br>Click to insert"
+COM_LOCALISE_TOOLTIP_TRANSLATION_EXTRA_KEYS_IN_TRANSLATION="If you do not want to keep a string in the translation, you should check the checkbox to remove it on 'Save'.<br> Some strings may not be in reference anymore.<br>Some strings may be specific plural forms that you need to keep."
```
maybe better use `select the checkbox`

infograf768 commented on administrator/components/com_localise/language/en-GB/com_localise.ini:
```diff
@@ -232,13 +240,20 @@ COM_LOCALISE_TEXT_TRANSLATION_TRANSLATED="Translated"
 COM_LOCALISE_TITLE_TRANSLATION="Translation"
-COM_LOCALISE_TOOLTIP_TRANSLATION_INSERT="<strong>Insert</strong><br>Click to insert"
+COM_LOCALISE_TOOLTIP_TRANSLATION_EXTRA_KEYS_IN_TRANSLATION="If you do not want to keep a string in the translation, you should check the checkbox to remove it on 'Save'.<br> Some strings may not be in reference anymore.<br>Some strings may be specific plural forms that you need to keep."
+COM_LOCALISE_TOOLTIP_TRANSLATION_EXTRA_KEYS_IN_TRANSLATION_ICON="Checking the box this string will be removed from your translation."
```
suggest
`"By checking the box this string will be removed from your translation."`

infograf768 commented on administrator/components/com_localise/language/en-GB/com_localise.ini:
```diff
@@ -232,13 +240,20 @@ COM_LOCALISE_TEXT_TRANSLATION_TRANSLATED="Translated"
+COM_LOCALISE_TOOLTIP_TRANSLATION_EXTRA_KEYS_IN_TRANSLATION="If you do not want to keep a string in the translation, you should check the checkbox to remove it on 'Save'.<br> Some strings may not be in reference anymore.<br>Some strings may be specific plural forms that you need to keep."
+COM_LOCALISE_TOOLTIP_TRANSLATION_EXTRA_KEYS_IN_TRANSLATION_ICON="Checking the box this string will be removed from your translation."
 COM_LOCALISE_TOOLTIP_TRANSLATION_GOOGLE="<strong>Google</strong><br>Click to find a google translation"
+COM_LOCALISE_TOOLTIP_TRANSLATION_GRAMMAR_CASE="Grammar case\\nIf it is a typing error in the language 'en-GB' if you check the box, this changed text will not be shown in the other 'xx-XX' languages."
```
I thought we would get rid of the  grammar case stuff and always use Modified values for en-GB when necessary. I'm confused.

Valc replied:
For me olny 2 options are posible on "modified values" with the "en-GB" language selected:

1. Deltete the check box from en-GB edition and keep the string as read only.
2. Use the check box for the only posible use: it is an en-GB grammar case than we can check as en-GB grammar case, also with the string as read-only.

Please, test it:
Go to en-GB edition and apply custom corrections on any string you can to edit there, due is allowed do it and "Save"
Then edit the same file in any other translation language and see the results of apply changes on en-GB reference language: all them are "fake" changes, "fake" changed text, etc.

Who will do it? No idea... but in this moment is allowed do it.

For me, all cases on normal edition at "en-GB"  view should be as "read-only" strings. if we access there we can revise, but not apply changes on our reference.

infograf768 commented on administrator/components/com_localise/language/en-GB/com_localise.ini:
```diff
@@ -232,13 +240,20 @@ COM_LOCALISE_TEXT_TRANSLATION_TRANSLATED="Translated"
+COM_LOCALISE_TOOLTIP_TRANSLATION_GRAMMAR_CASE="Grammar case\\nIf it is a typing error in the language 'en-GB' if you check the box, this changed text will not be shown in the other 'xx-XX' languages."
+COM_LOCALISE_TOOLTIP_TRANSLATION_INSERT="<strong>Insert</strong><br>Click to insert"
+COM_LOCALISE_TOOLTIP_TRANSLATION_KEY_TO_DELETE="Key to delete the next Joomla release"
+COM_LOCALISE_TOOLTIP_TRANSLATION_KEYS_TO_DELETE="What is shown below are the keys that are going to be removed in the next Joomla release, and not in this moment.<br>It is not recommended to edit or try to delete them because it would have the same effect as applying that action on your current 'en-GB' installed instance."
```
Suggest `The strings displayed below will be removed in the next Joomla release.<br>It is not recommended to edit or try to delete them because it would have the same effect as applying that action on your current 'en-GB' installed instance."`

Valc replied:
Done

infograf768 commented on administrator/components/com_localise/language/en-GB/com_localise.ini:
```diff
@@ -188,10 +188,13 @@ COM_LOCALISE_CANNOT_REMOVE_DEFAULT_LANGUAGE="Localise cannot remove default lang
 COM_LOCALISE_ACCESS_TRANSLATION_EDIT_DESC="Translate description"
+COM_LOCALISE_CHECKBOX_TRANSLATION_GRAMMAR_CASE="Grammar case"
 COM_LOCALISE_CONFIRM_TRANSLATION_SAVE="Before saving, make sure you have set if the translation is 'Complete' or not in the 'Details' tab."
+COM_LOCALISE_CONFIRM_TRANSLATION_SAVE_NOTINREF="Before saving, make sure to review if within the 'Not in ref' section you have selected the corresponding checkboxes of the strings to be deleted."
```
Suggest
`COM_LOCALISE_CONFIRM_TRANSLATION_SAVE_NOTINREF="Also, if you have strings in the 'Not in reference' section,  make sure you have selected the corresponding checkboxes of the strings you want to delete."`

Valc replied:
Done

Valc commented:
![comment](https://user-images.githubusercontent.com/417247/135912282-1eca2294-2980-410f-8c4b-441b63ca962a.png)

infograf768 commented:
@Valc 
before anything else:
`$has_notinref` is always equal to 0 in the view. Therefore we never get the 2 alerts.
also you forgot https://github.com/infograf768/j4localise/pull/35#discussion_r721181898

Valc commented:
Confirmed.
I will take a look asap. Job time now.

Valc commented:
@infograf768 
Seems that `$this->get('Item')->developdata` is always returning empty arrays from the data it is storing at that point, so, always "0".

Now is replaced by `$this->get('Item')->extra > 0` present at core or not core files, with or without develop.

Valc commented:
> I improved the "grammar case" tip display and corrected some strings:
> 
> <img alt="Screen Shot 2021-10-06 at 10 58 04" width="1453" src="https://user-images.githubusercontent.com/869724/136173938-953a42cb-8f9f-4781-9a3c-91118fc0a155.png">
> 
`

Yeah.. you are good :D "Ignore" matches prefectly there.


> Any reason to use `$this->get('Item')` instead of `$this->item` as we have `$this->item = $this->get('Item');` in the display method ?
> 
> it would give
> 
> ```
> $istranslation = $this->item->istranslation;
> $has_notinref  = $this->item->extra > 0 ? true : false;
> ``
I think is the same reply from both code lines. Not sure is "get" is due apply any "extra" filters to the data or not, new J4 format or not... really no idea, the result is the same.

infograf768 commented:
Merged. Thanks!
